src/video/cocoa/cocoa_v.mm

```diff
@@ -836,31 +566,187 @@ static void ClearWindowBuffer(uint32 *buffer, uint32 pitch, uint32 height)
 	if (this->buffer_depth == 8) {
 		free(this->pixel_buffer);
 		this->pixel_buffer = malloc(this->window_width * this->window_height);
-		if (this->pixel_buffer == NULL) {
-			DEBUG(driver, 0, "Failed to allocate pixel buffer");
-			return false;
-		}
+		if (this->pixel_buffer == nullptr) usererror("Out of memory allocating pixel buffer");
+	} else {
+		free(this->pixel_buffer);
+		this->pixel_buffer = nullptr;
 	}
 
-	static_cast<VideoDriver_Cocoa *>(VideoDriver::GetInstance())->GameSizeChanged();
-
 	/* Redraw screen */
 	this->num_dirty_rects = MAX_DIRTY_RECTS;
+	this->GameSizeChanged();
+}
+
+void VideoDriver_Cocoa::CheckPaletteAnim()
+{
+	if (_cur_palette.count_dirty != 0) {
+		Blitter *blitter = BlitterFactory::GetCurrentBlitter();
+
+		switch (blitter->UsePaletteAnimation()) {
+			case Blitter::PALETTE_ANIMATION_VIDEO_BACKEND:
+				this->UpdatePalette(_cur_palette.first_dirty, _cur_palette.count_dirty);
+				break;
+
+			case Blitter::PALETTE_ANIMATION_BLITTER:
+				blitter->PaletteAnimate(_cur_palette);
+				break;
+
+			case Blitter::PALETTE_ANIMATION_NONE:
+				break;
+
+			default:
+				NOT_REACHED();
+		}
+		_cur_palette.count_dirty = 0;
+	}
+}
+
+
+bool VideoDriver_Cocoa::PollEvent()
+{
+#ifdef _DEBUG
+	uint32 et0 = GetTick();
+#endif
+	NSEvent *event = [ NSApp nextEventMatchingMask:NSAnyEventMask
+				untilDate:[ NSDate distantPast ]
+				inMode:NSDefaultRunLoopMode dequeue:YES ];
+#ifdef _DEBUG
+	_tEvent += GetTick() - et0;
+#endif
+
+	if (event == nil) return false;
+
+	[ NSApp sendEvent:event ];
 
 	return true;
 }
 
 
-CocoaSubdriver *QZ_CreateWindowQuartzSubdriver(int width, int height, int bpp)
+void VideoDriver_Cocoa::GameLoop()
+{
+	uint32 cur_ticks = GetTick();
+	uint32 last_cur_ticks = cur_ticks;
+	uint32 next_tick = cur_ticks + MILLISECONDS_PER_TICK;
+
+#ifdef _DEBUG
+	uint32 et0 = GetTick();
+	uint32 st = 0;
+#endif
+
+	for (;;) {
+		@autoreleasepool {
+
+			uint32 prev_cur_ticks = cur_ticks; // to check for wrapping
+			InteractiveRandom(); // randomness
+
+			while (this->PollEvent()) {}
+
+			if (_exit_game) {
+				/* Restore saved resolution if in fullscreen mode. */
+				if (this->IsFullscreen()) _cur_resolution = this->orig_res;
+				break;
+			}
+
+			NSUInteger cur_mods = [ NSEvent modifierFlags ];
+
+#if defined(_DEBUG)
+			if (cur_mods & NSShiftKeyMask)
+#else
+			if (_tab_is_down)
+#endif
+			{
```

Comment:
This looks ugly, and made me think there was something going on with the if & elseif below. Stick the `{` in the preproc blocks?